Android/src/org/droidplanner/android/dialogs/UVCDialog.java

```diff
@@ -0,0 +1,209 @@
+        builder.setNegativeButton(android.R.string.cancel , mOnDialogClickListener);
+        builder.setNeutralButton(R.string.uvc_device_refresh, null);
+        final Dialog dialog = builder.create();
+        dialog.setCancelable(true);
+        dialog.setCanceledOnTouchOutside(true);
+        return dialog;
+    }
+
+    /**
+     * create view that this fragment shows
+     * @return
+     */
+    private final View initView() {
+        final View rootView = getActivity().getLayoutInflater().inflate(R.layout.dialog_uvc_device, null);
+        mSpinner = (Spinner)rootView.findViewById(R.id.spinner1);
+        final View empty = rootView.findViewById(android.R.id.empty);
+        mSpinner.setEmptyView(empty);
+        return rootView;
+    }
+
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        updateDevices();
+        final Button button = (Button)getDialog().findViewById(android.R.id.button3);
+        if (button != null) {
+            button.setOnClickListener(mOnClickListener);
+        }
+    }
+
+    private final View.OnClickListener mOnClickListener = new View.OnClickListener() {
+        @Override
+        public void onClick(final View v) {
+            switch (v.getId()) {
+                case android.R.id.button3:
+                    updateDevices();
+                    break;
+            }
+        }
+    };
+
+    private final DialogInterface.OnClickListener mOnDialogClickListener = new DialogInterface.OnClickListener() {
+        @Override
+        public void onClick(final DialogInterface dialog, final int which) {
+            switch (which) {
+                case DialogInterface.BUTTON_POSITIVE:
+                    final Object item = mSpinner.getSelectedItem();
+                    if (item instanceof UsbDevice) {
+                        mUSBMonitor.requestPermission((UsbDevice)item);
+                    }
+                    break;
+            }
+        }
+    };
+
+    public void updateDevices() {
+        mUSBMonitor.dumpDevices();
+        final List<DeviceFilter> filter = DeviceFilter.getDeviceFilters(getActivity(), R.xml.uvc_device_filter);
+        mDeviceListAdapter = new DeviceListAdapter(getActivity(), mUSBMonitor.getDeviceList(filter.get(0)));
+        mSpinner.setAdapter(mDeviceListAdapter);
+    }
+
+    private static final class DeviceListAdapter extends BaseAdapter {
+
+        private final LayoutInflater mInflater;
+        private final List<UsbDevice> mList;
+
+        public DeviceListAdapter(final Context context, final List<UsbDevice>list) {
+            mInflater = LayoutInflater.from(context);
+            mList = list != null ? list : new ArrayList<UsbDevice>();
+        }
+
+        @Override
+        public int getCount() {
+            return mList.size();
+        }
+
+        @Override
+        public UsbDevice getItem(final int position) {
+            if ((position >= 0) && (position < mList.size()))
```

Comment:
@guiseco you don't need to check the validity of the `position` argument here, since it's based on the `getCount()` that you overrode above.